src/dune_rules/dune_file.ml

```diff
@@ -548,6 +557,50 @@ module Library = struct
     let field = field_o "wrapped" (located decode)
   end
 
+  let library_mode_field ~stanza_loc ~dune_version project =
+    let default = Mode_conf.Lib.Set.default stanza_loc in
+    let ordered_mode_field ~stanza_loc project =
+      let+ modes = field_o "modes" Ordered_set_lang.decode in
+      match modes with
+      | None -> default
+      | Some modes ->
+        let modes =
+          Ordered_set_lang.eval modes
+            ~parse:(fun ~loc s ->
+              let mode =
+                Dune_lang.Decoder.parse Mode_conf.Lib.decode
+                  (Dune_project.parsing_context project)
+                  (Atom (loc, Dune_lang.Atom.of_string s))
+              in
+              (mode, Mode_conf.Kind.Requested loc))
+            ~eq:(fun (a, _) (b, _) -> Mode_conf.Lib.equal a b)
+            ~standard:[ (Ocaml Best, Mode_conf.Kind.Requested stanza_loc) ]
+        in
+        Mode_conf.Lib.Set.of_list modes
+    in
+    let expected_version = (3, 8) in
+    if dune_version >= expected_version then
+      ordered_mode_field ~stanza_loc project
+    else
+      (* Old behavior: if old parser succeeds, return that. Otherwise, if
+         parsing the ordered set language succeeds, ask the user to upgrade to
+         a supported version. Otherwise, fail with the first error. *)
+      try_
+        (field "modes"
+           (Mode_conf.Lib.Set.decode >>| fun modes -> `Old modes)
+           ~default:(`Old default))
+        (fun exn ->
+          try_
+            ( ordered_mode_field ~stanza_loc project >>| fun modes ->
+              if dune_version >= expected_version then `New modes else `Upgrade
```

Comment:
this might be too complicated for its own good, but I wanted to return the first error message if we fail parsing the OSL for old versions of the dune language.